src/Security/AuthenticationRequestFilter.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace SilverStripe\Security;
+
+use SilverStripe\Control\HTTPResponse_Exception;
+use SilverStripe\Control\RequestFilter;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Control\Session;
+use SilverStripe\ORM\DataModel;
+use SilverStripe\Core\Config\Configurable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\ORM\ValidationException;
+
+class AuthenticationRequestFilter implements RequestFilter, IdentityStore
+{
+
+    use Configurable;
+
+    /**
+     * @return array|IdentityStore[]
+     */
+    protected function getHandlers()
+    {
+        return array_map(
+            function ($identifier) {
+                return Injector::inst()->get($identifier);
+            },
+            static::config()->get('handlers')
+        );
+    }
+
+    /**
+     * Identify the current user from the request
+     *
+     * @param HTTPRequest $request
+     * @param Session $session
+     * @param DataModel $model
+     * @return bool|void
+     * @throws HTTPResponse_Exception
+     */
+    public function preRequest(HTTPRequest $request, Session $session, DataModel $model)
+    {
+        try {
+            /** @var AuthenticationHandler $handler */
+            foreach ($this->getHandlers() as $handler) {
+                // @todo Update requestfilter logic to allow modification of initial response
+                // in order to add cookies, etc
+                $member = $handler->authenticateRequest($request, new HTTPResponse());
+                if ($member) {
+                    // @todo Remove the static coupling here
+                    Security::setCurrentUser($member);
+                    break;
+                }
+            }
+        } catch (ValidationException $e) {
```

Comment:
tractorcow 2 days ago Contributor

We should probably catch only ValidationException; if the exception is already a HTTPResponse_Exception it should bubble up automatically.

This also allows us to get the validation error for the response.
Firesphere 2 days ago Contributor

ValidationException is currently an ORM exception, hence the comment. Happy to switch :)